harp/clock.py

```diff
@@ -0,0 +1,220 @@
+        When inputting timestamps in seconds, use a sample rate of 1
+    baud_rate : float
+        The baud rate of the clock signal
+
+    Returns
+    -------
+    starts : np.array
+        Sample numbers or timestamps at which each second begins
+    harp_times : np.array
+        Harp clock times in seconds
+    """
+
+    min_delta = sample_rate / 2  # 0.5 seconds
+
+    barcode_edges = get_barcode_edges(timestamps_or_sample_numbers, min_delta)
+
+    start_samples = [timestamps_or_sample_numbers[edges[0]] for edges in barcode_edges]
+
+    harp_times = [
+        convert_barcode(
+            timestamps_or_sample_numbers[edges[0] : edges[1]] / sample_rate,
+            states[edges[0] : edges[1]],
+            baud_rate=baud_rate,
+        )
+        for edges in barcode_edges
+    ]
+
+    harp_times_corrected = correct_outliers(np.array(harp_times))
+
+    return np.array(start_samples), harp_times_corrected
+
+
+def get_barcode_edges(sample_numbers, min_delta):
+    """
+    Returns the start and end indices of each barcode
+
+    Parameters
+    ----------
+    sample_numbers : np.array
+        Sample numbers of clock line (high and low states)
+    min_delta : int
+        The minimum length between the end of one
+        barcode and the start of the next
+
+    Returns
+    -------
+    edges : list of tuples
+        Contains the start and end indices of each barcode
+    """
+
+    (splits,) = np.where(np.diff(sample_numbers) > min_delta)
+
+    return list(zip(splits[:-1] + 1, splits[1:] + 1))
+
+
+def convert_barcode(transition_times, states, baud_rate):
+    """
+    Converts Harp clock barcode to a clock time in seconds
+
+    Parameters
+    ----------
+    transition_times : np.array
+        Times (in seconds) each clock line transition
+    states : np.array
+        states (1 or 0) for each clock line transition
+    baud_rate : float
+        The baud rate of the clock signal
+
+    Returns
+    -------
+    harp_time : int
+        Harp time in seconds for the current barcode
+
+    """
+
+    intervals = np.round(np.diff(transition_times * baud_rate)).astype("int")
+
+    barcode = np.concatenate(
+        [np.ones((count,)) * state for state, count in zip(states[:-1], intervals)]
+    ).astype("int")
+
+    val = np.concatenate(
+        (np.arange(1, 9), np.arange(11, 19), np.arange(21, 29), np.arange(31, 39))
+    )
+
+    s = np.flip(barcode[val])
+    harp_time = s.dot(2 ** np.arange(s.size)[::-1])
+
+    return harp_time
+
+
+def correct_outliers(harp_times):
+    """
+    Corrects outliers in the Harp clock times
+
+    Parameters
+    ----------
+    harp_times : np.array
+        Harp clock times in seconds
+
+    Returns
+    -------
+    corrected_harp_times : np.array
+        Corrected Harp clock times in seconds
+    """
+
+    diffs = np.abs(np.diff(harp_times))
+
+    # Find the indices of the outliers
+    outliers = np.where(diffs > 1)[0]
```

Comment:
Isn't this to strict? I guess it's possible that the harp period is slightly over 1s

One possibility could be adding a `delta_eps=0.01` argument and use `np.where(diffs > 1 + delta_eps)[0]`

Reply:
The Harp times will always be exactly 1 s apart, this is programmed into the firmware. 

Right now this approach is not robust to the Harp clock resetting, or gaps in the ephys recording. A better solution might be to look for any Harp times with outliers both before *and* after, and remove these entirely. That will make it possible to detect jumps in the Harp timestamps that are actually valid.